# Limpeza dos dados
    Neste notebook, vamos preparar o conjunto de dados para nossa análise futura. Abaixo está o plano cronológico das etapas:
- 1. Importando as Biblíotecas
- 2. Obtendo informações sobre a dataframe
     - 2.1 Visualização inicial
     - 2.2 Contagem dos valores nulos e duplicatas
- 3. Limpeza de Dados e Tratamento
     - 3.1 Remoção de Colunas
     - 3.2 Eliminação dos Prefixos / Alterando nomes das colunas
     - 3.3 Atribuindo os Valores
     - 3.4 Manipulação para a memória
- 4. Exportando

## 1. Importando as biblíotecas

In [3]:

import pandas as pd
import numpy as np
%config IPCompleter.greedy=True
pd.set_option("display.max_columns", None)


## 2. Obtaining informations about our dataframe
- Nesta seção nós vamos ler a nossa dataframe, conhecer o formato e o tipo de data de cada coluna, e análisar quantas duplicátas e valores nulos há nessa dataframe.



### 2.1 Visualização ínicial

In [6]:
t = pd.read_csv(r"C:\Users\Lucas\Data Science P1\Data Science\DADOS\MICRODADOS.csv",delimiter =';',encoding="latin-1")

In [7]:
print(f'Nós temos {t.shape[0]} índices e {t.shape[1]} colunas na dataframe')

Nós temos 3476105 índices e 76 colunas na dataframe


In [8]:
t.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3476105 entries, 0 to 3476104
Data columns (total 76 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   NU_INSCRICAO            int64  
 1   NU_ANO                  int64  
 2   TP_FAIXA_ETARIA         int64  
 3   TP_SEXO                 object 
 4   TP_ESTADO_CIVIL         int64  
 5   TP_COR_RACA             int64  
 6   TP_NACIONALIDADE        int64  
 7   TP_ST_CONCLUSAO         int64  
 8   TP_ANO_CONCLUIU         int64  
 9   TP_ESCOLA               int64  
 10  TP_ENSINO               float64
 11  IN_TREINEIRO            int64  
 12  CO_MUNICIPIO_ESC        float64
 13  NO_MUNICIPIO_ESC        object 
 14  CO_UF_ESC               float64
 15  SG_UF_ESC               object 
 16  TP_DEPENDENCIA_ADM_ESC  float64
 17  TP_LOCALIZACAO_ESC      float64
 18  TP_SIT_FUNC_ESC         float64
 19  CO_MUNICIPIO_PROVA      int64  
 20  NO_MUNICIPIO_PROVA      object 
 21  CO_UF_PROVA             int64  

In [9]:
t.describe()

NU_INSCRICAO     NU_ANO  TP_FAIXA_ETARIA  TP_ESTADO_CIVIL  \
count  3.476105e+06  3476105.0     3.476105e+06     3.476105e+06   
mean   2.100563e+11     2022.0     4.863503e+00     1.042771e+00   
std    1.023449e+06        0.0     3.690560e+00     3.873107e-01   
min    2.100545e+11     2022.0     1.000000e+00     0.000000e+00   
25%    2.100554e+11     2022.0     2.000000e+00     1.000000e+00   
50%    2.100563e+11     2022.0     3.000000e+00     1.000000e+00   
75%    2.100571e+11     2022.0     6.000000e+00     1.000000e+00   
max    2.100580e+11     2022.0     2.000000e+01     4.000000e+00   

        TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  \
count  3.476105e+06      3.476105e+06     3.476105e+06     3.476105e+06   
mean   2.035284e+00      1.028641e+00     1.684877e+00     2.178502e+00   
std    1.019759e+00      2.102919e-01     7.297286e-01     4.064238e+00   
min    0.000000e+00      0.000000e+00     1.000000e+00     0.000000e+00   
25%    1.000000e+00      1.000000e+00     1.000000e+00     0.000000e+00   
50%    2.000000e+00      1.000000e+00     2.000000e+00     0.000000e+00   
75%    3.000000e+00      1.000000e+00     2.000000e+00     2.000000e+00   
max    5.000000e+00      4.000000e+00     4.000000e+00     1.600000e+01   

          TP_ESCOLA     TP_ENSINO  IN_TREINEIRO  CO_MUNICIPIO_ESC  \
count  3.476105e+06  1.262744e+06  3.476105e+06      9.519440e+05   
mean   1.440080e+00  1.005993e+00  1.475019e-01      3.201401e+06   
std    6.070447e-01  7.717901e-02  3.546056e-01      1.024569e+06   
min    1.000000e+00  1.000000e+00  0.000000e+00      1.100015e+06   
25%    1.000000e+00  1.000000e+00  0.000000e+00      2.401008e+06   
50%    1.000000e+00  1.000000e+00  0.000000e+00      3.205309e+06   
75%    2.000000e+00  1.000000e+00  0.000000e+00      3.550308e+06   
max    3.000000e+00  2.000000e+00  1.000000e+00      5.300108e+06   

           CO_UF_ESC  TP_DEPENDENCIA_ADM_ESC  TP_LOCALIZACAO_ESC  \
count  951944.000000           951944.000000       951944.000000   
mean       31.865961                2.435363            1.035228   
std        10.220105                0.909683            0.184355   
min        11.000000                1.000000            1.000000   
25%        24.000000                2.000000            1.000000   
50%        32.000000                2.000000            1.000000   
75%        35.000000                2.000000            1.000000   
max        53.000000                4.000000            2.000000   

       TP_SIT_FUNC_ESC  CO_MUNICIPIO_PROVA   CO_UF_PROVA  TP_PRESENCA_CN  \
count    951944.000000        3.476105e+06  3.476105e+06    3.476105e+06   
mean          1.015350        3.106623e+06  3.092124e+01    6.785034e-01   
std           0.197644        1.010732e+06  1.007864e+01    4.680213e-01   
min           1.000000        1.100015e+06  1.100000e+01    0.000000e+00   
25%           1.000000        2.405801e+06  2.400000e+01    0.000000e+00   
50%           1.000000        3.127701e+06  3.100000e+01    1.000000e+00   
75%           1.000000        3.550308e+06  3.500000e+01    1.000000e+00   
max           4.000000        5.300108e+06  5.300000e+01    2.000000e+00   

       TP_PRESENCA_CH  TP_PRESENCA_LC  TP_PRESENCA_MT   CO_PROVA_CN  \
count    3.476105e+06    3.476105e+06    3.476105e+06  2.355395e+06   
mean     7.200030e-01    7.200030e-01    6.785034e-01  1.087859e+03   
std      4.519873e-01    4.519873e-01    4.680213e-01  1.166275e+01   
min      0.000000e+00    0.000000e+00    0.000000e+00  1.085000e+03   
25%      0.000000e+00    0.000000e+00    0.000000e+00  1.086000e+03   
50%      1.000000e+00    1.000000e+00    1.000000e+00  1.087000e+03   
75%      1.000000e+00    1.000000e+00    1.000000e+00  1.088000e+03   
max      2.000000e+00    2.000000e+00    2.000000e+00  1.190000e+03   

        CO_PROVA_CH   CO_PROVA_LC   CO_PROVA_MT    NU_NOTA_CN    NU_NOTA_CH  \
count  2.493442e+06  2.493442e+06  2.355395e+06  2.355395e+06  2.493442e+06   


In [10]:
t.head(5)

NU_INSCRICAO  NU_ANO  TP_FAIXA_ETARIA TP_SEXO  TP_ESTADO_CIVIL  \
0  210057943671    2022               14       M                2   
1  210057516120    2022               14       M                2   
2  210057280536    2022                5       F                1   
3  210055724397    2022                6       M                1   
4  210055097896    2022                4       M                0   

   TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  TP_ESCOLA  \
0            2                 1                1                2          1   
1            1                 1                1               16          1   
2            2                 1                1                2          1   
3            3                 1                1                2          1   
4            3                 1                1                1          1   

   TP_ENSINO  IN_TREINEIRO  CO_MUNICIPIO_ESC NO_MUNICIPIO_ESC  CO_UF_ESC  \
0        NaN             0               NaN              NaN        NaN   
1        NaN             0               NaN              NaN        NaN   
2        NaN             0               NaN              NaN        NaN   
3        NaN             0               NaN              NaN        NaN   
4        NaN             0               NaN              NaN        NaN   

  SG_UF_ESC  TP_DEPENDENCIA_ADM_ESC  TP_LOCALIZACAO_ESC  TP_SIT_FUNC_ESC  \
0       NaN                     NaN                 NaN              NaN   
1       NaN                     NaN                 NaN              NaN   
2       NaN                     NaN                 NaN              NaN   
3       NaN                     NaN                 NaN              NaN   
4       NaN                     NaN                 NaN              NaN   

   CO_MUNICIPIO_PROVA         NO_MUNICIPIO_PROVA  CO_UF_PROVA SG_UF_PROVA  \
0             5300108                   Brasília           53          DF   
1             5300108                   Brasília           53          DF   
2             2925758  Presidente Tancredo Neves           29          BA   
3             3201308                  Cariacica           32          ES   
4             1506203                Salinópolis           15          PA   

   TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  TP_PRESENCA_MT  \
0               0               0               0               0   
1               0               0               0               0   
2               1               1               1               1   
3               1               1               1               1   
4               0               0               0               0   

   CO_PROVA_CN  CO_PROVA_CH  CO_PROVA_LC  CO_PROVA_MT  NU_NOTA_CN  NU_NOTA_CH  \
0          NaN          NaN          NaN          NaN         NaN         NaN   
1          NaN          NaN          NaN          NaN         NaN         NaN   
2       1087.0       1056.0       1066.0       1078.0       421.1       546.0   
3       1088.0       1058.0       1067.0       1077.0       490.7       388.6   
4          NaN          NaN          NaN          NaN         NaN         NaN   

   NU_NOTA_LC  NU_NOTA_MT                                TX_RESPOSTAS_CN  \
0         NaN         NaN                                            NaN   
1         NaN         NaN                                            NaN   
2       498.8       565.3  DCDCEBADDABEDBCBEAEACCDDECEEBECBECAAECDDDDBCE   
3       357.8       416.0  BACBBABBCAABCABBCBACBCBCBACABACABCBDBCBABAACA   
4         NaN         NaN                                            NaN   

                                 TX_RESPOSTAS_CH  \
0                                            NaN   
1                                            NaN   
2  DBDABCADADECACBDCCEDACCCECDDABDCEADABCBCBAEAB   
3  BBCAACACDCABCBACCABABCBACBCABBBBBABCBACCBCBBA   
4                                            NaN   

                                 TX_RESPOSTAS_LC  \


Após a análise inicial do nosso dataframe, já percebemos que ele é extenso, contendo mais de 3 milhões de linhas e mais de 70 colunas. A variedade de dados presentes nos permitirá realizar uma análise robusta para identificar os principais fatores que impactam a nota geral dos candidatos. Algumas observações importantes após essa primeira análise incluem:

- **Prefixos desnecessários nas colunas**: Identificamos a presença de prefixos como TP_, NU_, ST_ em nossas colunas, que podem ser removidos para facilitar a leitura.
- **Valores nulos**: Notamos que algumas colunas possuem uma quantidade significativa de valores nulos, o que exigirá tratamento adequado.
- **Otimização de tipos de dados**: Há colunas atualmente classificadas como int64, que consomem muita memória. Podemos substituir esses tipos por int16 para economizar espaço sem perder a integridade dos dados.
- **Substituição por categorias**: Com base nas informações disponíveis nos dicionários de dados, algumas sequências numéricas podem ser substituídas por categorias, tornando a análise mais intuitiva. Por exemplo, a coluna TP_FAIXA_ETARIA pode ter valores numéricos convertidos em categorias mais descritivas.
- **Seleção de colunas relevantes**: Nem todas as colunas são essenciais para nossa análise. Em etapas futuras, discutiremos quais colunas poderão ser descartadas.


### 2.2 Contagem dos valores nulos e duplicatas

In [13]:
nan = t.apply([lambda x: (round(x.isna().sum()/(x.count() + x.isna().sum()),3)) * 100]).T
nan.columns = ['NaN %']
nan[nan['NaN %'] >= 70]


NaN %
CO_MUNICIPIO_ESC         72.6
NO_MUNICIPIO_ESC         72.6
CO_UF_ESC                72.6
SG_UF_ESC                72.6
TP_DEPENDENCIA_ADM_ESC   72.6
TP_LOCALIZACAO_ESC       72.6
TP_SIT_FUNC_ESC          72.6

In [14]:
f'A nossa dataframe tem {t.duplicated().sum()} duplicatas'


'A nossa dataframe tem 0 duplicatas'

Colunas com mais de 70% de valores nulos deverão ser removidas, pois prejudicarão nossa análise futura e consumirão memória desnecessariamente.

## 3. Limpeza e tratamento de dados
    Na etapa de tratamento de dados, realizaremos as seguintes ações:
- Remoção de prefixos desnecessários: Limparemos as colunas, eliminando prefixos que não acrescentam valor à análise.

- Eliminação de colunas irrelevantes: Identificaremos e retiraremos colunas que não são pertinentes, focando apenas nas informações cruciais para nossa análise.

- Feature Engineering: Transformaremos dados brutos em categorias mais eficazes, facilitando a análise e aumentando a precisão dos nossos insights.

- Otimização da memória: Ajustaremos os tipos de dados para economizar memória e melhorar a eficiência do processamento.


### 3.1 Remoção de Colunas
Dentre algumas colunas que iremos resolver, podemos mencionar algumas:
 - **NU_INSCRICAO**:  Esta coluna contém o número de ID da inscrição do aluno. Como essa informação não contribui para a nossa análise e apenas consome memória, será removida.
 - **NU_ANO**:O ano da prova já é conhecido anteriormente, tornando esta coluna redundante para nossa análise.
 - **CO_MUNICIPIO_PROVA**: Em vez de analisar os dados a nível municipal, é mais relevante e prático focar nas unidades federativas (estados), o que simplifica nossa análise.
 - **TX_GABARITO_CN,TX_GABARITO_CH,TX_GABARITO_LC,TX_GABARITO_MT**: O Exame Nacional do Ensino Médio (ENEM) utiliza a Teoria de Resposta ao Item (TRI) para calcular as notas finais dos candidatos. Portanto, as colunas que contêm as respostas dos alunos e seus gabaritos não são necessárias, pois não conseguimos quantificar os acertos diretamente.
 - **CO_MUNICIPIO_ESC,NO_MUNICIPIO_ESC,CO_UF_ESC,SG_UF_ESC,TP_DEPENDENCIA_ADM_ESC,TP_LOCALIZACAO_ESC,TP_SIT_FUNC_ESC** = Estas colunas possuem um excesso de valores nulos, o que prejudica nossa análise e consome recursos de memória de forma desnecessária.

Os significados e detalhes de cada coluna podem ser consultados no dicionário de dados, que será uma ferramenta útil ao longo do processo de análise.



In [18]:
columnsdrop = t.iloc[:,[*range(0,2),6,7,8,10,*range(12,15),*range(16,24),*range(24,31),*range(35,50),*range(53,56),*range(57,72),73]]

In [19]:
columnsdrop.head(2)

NU_INSCRICAO  NU_ANO  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  \
0  210057943671    2022                 1                1                2   
1  210057516120    2022                 1                1               16   

   TP_ENSINO  CO_MUNICIPIO_ESC NO_MUNICIPIO_ESC  CO_UF_ESC  \
0        NaN               NaN              NaN        NaN   
1        NaN               NaN              NaN        NaN   

   TP_DEPENDENCIA_ADM_ESC  TP_LOCALIZACAO_ESC  TP_SIT_FUNC_ESC  \
0                     NaN                 NaN              NaN   
1                     NaN                 NaN              NaN   

   CO_MUNICIPIO_PROVA NO_MUNICIPIO_PROVA  CO_UF_PROVA SG_UF_PROVA  \
0             5300108           Brasília           53          DF   
1             5300108           Brasília           53          DF   

   TP_PRESENCA_CN  TP_PRESENCA_CH  TP_PRESENCA_LC  TP_PRESENCA_MT  \
0               0               0               0               0   
1               0               0               0               0   

   CO_PROVA_CN  CO_PROVA_CH  CO_PROVA_LC  CO_PROVA_MT TX_RESPOSTAS_CN  \
0          NaN          NaN          NaN          NaN             NaN   
1          NaN          NaN          NaN          NaN             NaN   

  TX_RESPOSTAS_CH TX_RESPOSTAS_LC TX_RESPOSTAS_MT  TP_LINGUA TX_GABARITO_CN  \
0             NaN             NaN             NaN          0            NaN   
1             NaN             NaN             NaN          0            NaN   

  TX_GABARITO_CH TX_GABARITO_LC TX_GABARITO_MT  TP_STATUS_REDACAO  \
0            NaN            NaN            NaN                NaN   
1            NaN            NaN            NaN                NaN   

   NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  NU_NOTA_COMP4  NU_NOTA_COMP5  \
0            NaN            NaN            NaN            NaN            NaN   
1            NaN            NaN            NaN            NaN            NaN   

  Q003 Q004  Q005 Q007 Q008 Q009 Q010 Q011 Q012 Q013 Q014 Q015 Q016 Q017 Q018  \
0    A    B     3    B    B    B    B    B    B    A    B    B    B    A    A   
1    E    E     3    D    E    E    E    E    E    E    E    E    E    E    B   

  Q019 Q020 Q021 Q023  
0    A    A    A    A  
1    E    B    B    B

In [20]:
t.drop(columnsdrop,axis=1,inplace=True)

In [21]:
t.head(5)

TP_FAIXA_ETARIA TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_ESCOLA  \
0               14       M                2            2          1   
1               14       M                2            1          1   
2                5       F                1            2          1   
3                6       M                1            3          1   
4                4       M                0            3          1   

   IN_TREINEIRO SG_UF_ESC  NU_NOTA_CN  NU_NOTA_CH  NU_NOTA_LC  NU_NOTA_MT  \
0             0       NaN         NaN         NaN         NaN         NaN   
1             0       NaN         NaN         NaN         NaN         NaN   
2             0       NaN       421.1       546.0       498.8       565.3   
3             0       NaN       490.7       388.6       357.8       416.0   
4             0       NaN         NaN         NaN         NaN         NaN   

   NU_NOTA_REDACAO Q001 Q002 Q006 Q022 Q024 Q025  
0              NaN    A    A    B    A    A    A  
1              NaN    D    D    Q    E    E    B  
2            760.0    E    F    B    C    A    B  
3            320.0    C    A    A    C    B    B  
4              NaN    D    B    B    B    A    A

In [22]:
f'Nós continuamos com o mesmo valores de linhas, entretanto tiramos {columnsdrop.shape[1]} colunas, agora temos {t.shape[1]}'

'Nós continuamos com o mesmo valores de linhas, entretanto tiramos 58 colunas, agora temos 18'

### 3.2 Eliminação dos Prefixos / Alterando nomes das colunas
- **Prefixos:** ST,TP,IN,NU,SG

In [24]:
t.head(1)

TP_FAIXA_ETARIA TP_SEXO  TP_ESTADO_CIVIL  TP_COR_RACA  TP_ESCOLA  \
0               14       M                2            2          1   

   IN_TREINEIRO SG_UF_ESC  NU_NOTA_CN  NU_NOTA_CH  NU_NOTA_LC  NU_NOTA_MT  \
0             0       NaN         NaN         NaN         NaN         NaN   

   NU_NOTA_REDACAO Q001 Q002 Q006 Q022 Q024 Q025  
0              NaN    A    A    B    A    A    A

In [25]:
t.columns

Index(['TP_FAIXA_ETARIA', 'TP_SEXO', 'TP_ESTADO_CIVIL', 'TP_COR_RACA',
       'TP_ESCOLA', 'IN_TREINEIRO', 'SG_UF_ESC', 'NU_NOTA_CN', 'NU_NOTA_CH',
       'NU_NOTA_LC', 'NU_NOTA_MT', 'NU_NOTA_REDACAO', 'Q001', 'Q002', 'Q006',
       'Q022', 'Q024', 'Q025'],
      dtype='object')

In [26]:
t.columns = t.columns.str.replace(r"\b(TP_|IN_|NU_|SG_|ST)","",regex=True).str.capitalize()
## Usando regex para retirar os prefixos e usando o metodo capitalize para apenas a primeira parte ter a letra maiúscula

In [27]:
Questions = {'Q001':'Escolaridade_pai','Q002':'Escolaridade_mae','Q006':'Renda_familiar','Q022':'Celular','Q024':'Computador','Q025':'Internet'}
## Visualização no dicionário o nome das variáveis

In [28]:
t.rename(Questions,axis=1,inplace=True)

In [29]:
t.columns

Index(['Faixa_etaria', 'Sexo', 'Estado_civil', 'Cor_raca', 'Escola',
       'Treineiro', 'Uf_esc', 'Nota_cn', 'Nota_ch', 'Nota_lc', 'Nota_mt',
       'Nota_redacao', 'Escolaridade_pai', 'Escolaridade_mae',
       'Renda_familiar', 'Celular', 'Computador', 'Internet'],
      dtype='object')

In [30]:
t.sample(1)

Faixa_etaria Sexo  Estado_civil  Cor_raca  Escola  Treineiro Uf_esc  \
2426942             4    F             1         1       1          0    NaN   

         Nota_cn  Nota_ch  Nota_lc  Nota_mt  Nota_redacao Escolaridade_pai  \
2426942    535.9    574.4    562.3    709.2         860.0                F   

        Escolaridade_mae Renda_familiar Celular Computador Internet  
2426942                G              G       E          C        B

### 3.3 Atribuindo os valores

In [32]:

##Categorizando as idades
Answers_Age = {(1,2,3,4):('<=19, Adolescente'),
               (5,6,7,8,9):('20 - 25 , Jovens Adultos'),
               (10,11,12,13,14):('26 - 44, Adulto'),
               (15,16,17,18):('45 - 65, Meia-idade'),
               (19,20):('<66, Idosos')}
##Categorizando o Estado Cívil
Answer_CivilState= {0:'Não Informado',
                    1:'Solteiro(a)',
                    2:'Casado(a)',
                    3:'Divorciado(a)',
                    4:'Viúvo(a)'}
##Categorizando a raça
Answer_Race = {0:'Não declarado',
               1:'Branca',
               2:'Preta',
               3:'Parda',
               4:'Amarela',
               5:'Indigena',
               6:'Não dispõe da informação'}
##Categorizando tipo de escola
Answer_School = {1:'Não Respondeu',
                 2:'Pública',
                 3:'Privada'}
##Categorizando escolaridade pai,mãe
Answer_Q012 = {"A":'Nunca Estudou',
               "B":'Educação Infantil',
               "C":'Fundamental 1',
               "D":'Fundamental 2',
               "E":'Ensino Médio',
               "F":'Graduação Universitária',
               "G":'Pós graduação',
               "H":'Não sei'}
##Categorizando numero de celulares,computadores
Answer_CellComp = {"A":"Não",
                   "B":"Sim, um",
                   "C":"Sim, dois",
                   "D":"Sim, três",
                   "E":"Sim, quatro ou mais"}
##Categorizando o  acesso a internet
Answer_Internet = {"A":"Não",
                   "B":"Sim"}
## Identificação de Treineiro
Answer_T = {0:"Não",
            1:"Sim"}
##Categorizando a renda famíliar
Answer_Income = {'A':'Nenhuma Renda',
                 'B':'Ate 1.212,00',
                 'C':'De 1.212,01 ate 1.818,00',
                 'D':'De 1.818,01 ate 2.424,00',
                 'E':'De 2.424,01 ate 3.030,00',
                 'F':'De 3.030,01 ate 3.636,00',
                 'G':'De 3.636,01 ate 4.848,00',
                 'H':'De 4.848,01 ate 6.060,00',
                 'I':'De 6.060,01 ate 7.272,00',
                 'J':'De 7.272,01 ate 8.484,00',
                 'K':'De 8.484,01 ate 9.696,00',
                 'L':'De 9.696,01 ate 10.908,00',
                 'M':'De 10.908,01 ate 12.120,00',
                 'N':'De 12.120,01 ate 14.544,00',
                 'O':'De 14.544,01 ate 18.180,00',
                 'P':'De 18.180,01 ate 24.240,00',
                 'Q':'Acima de 24.240,00'}

In [33]:
t['Nota_geral'] = (t['Nota_cn'] + t['Nota_ch'] + t['Nota_lc'] + t['Nota_mt'] + t['Nota_redacao'])/5
## Criação do valor da nota geral, que será a média das outras notas

In [34]:
## Atribuindo os valores a nossas colunas
t['Faixa_etaria'] = t['Faixa_etaria'].replace(Answers_Age)
t['Estado_civil'] = t['Estado_civil'].replace(Answer_CivilState)
t['Cor_raca'] = t['Cor_raca'].replace(Answer_Race)
t['Escola'] = t['Escola'].replace(Answer_School)
t['Escolaridade_pai'] = t['Escolaridade_pai'].replace(Answer_Q012)
t['Escolaridade_mae'] = t['Escolaridade_mae'].replace(Answer_Q012)
t['Celular'] = t['Celular'].replace(Answer_CellComp)
t['Computador'] = t['Computador'].replace(Answer_CellComp)
t['Internet'] = t['Internet'].replace(Answer_Internet)
t['Treineiro'] = t['Treineiro'].replace(Answer_T)
t['Renda_familiar'] = t['Renda_familiar'].replace(Answer_Income)

In [35]:
t.sample(10)

Faixa_etaria Sexo   Estado_civil Cor_raca         Escola  \
1043297         <=19, Adolescente    F    Solteiro(a)   Branca        Pública   
2355905         <=19, Adolescente    F    Solteiro(a)   Branca        Privada   
2764872  20 - 25 , Jovens Adultos    M    Solteiro(a)    Preta  Não Respondeu   
1713255  20 - 25 , Jovens Adultos    M    Solteiro(a)    Parda  Não Respondeu   
3038365         <=19, Adolescente    M    Solteiro(a)    Parda  Não Respondeu   
844591            26 - 44, Adulto    M    Solteiro(a)    Preta  Não Respondeu   
2330555         <=19, Adolescente    F    Solteiro(a)   Branca  Não Respondeu   
2419891         <=19, Adolescente    F  Não Informado   Branca  Não Respondeu   
1813163  20 - 25 , Jovens Adultos    M    Solteiro(a)   Branca  Não Respondeu   
597158   20 - 25 , Jovens Adultos    F    Solteiro(a)    Parda  Não Respondeu   

        Treineiro Uf_esc  Nota_cn  Nota_ch  Nota_lc  Nota_mt  Nota_redacao  \
1043297       Não    NaN      NaN      NaN      NaN      NaN           NaN   
2355905       Não     SE    549.2    580.1    602.6    659.8         860.0   
2764872       Não    NaN      NaN      NaN      NaN      NaN           NaN   
1713255       Não    NaN    443.8    389.9    441.4    627.6         360.0   
3038365       Não    NaN    544.3    589.9    562.8    557.3         620.0   
844591        Não    NaN      NaN      NaN      NaN      NaN           NaN   
2330555       Sim    NaN    410.6    521.2    517.5    571.3         540.0   
2419891       Não    NaN    416.1    498.7    469.6    510.2         540.0   
1813163       Não    NaN    514.6    518.4    451.0    563.8         600.0   
597158        Não    NaN      NaN      NaN      NaN      NaN           NaN   

                Escolaridade_pai         Escolaridade_mae  \
1043297            Fundamental 2             Ensino Médio   
2355905  Graduação Universitária  Graduação Universitária   
2764872            Fundamental 1            Fundamental 1   
1713255        Educação Infantil            Fundamental 2   
3038365        Educação Infantil             Ensino Médio   
844591              Ensino Médio            Fundamental 1   
2330555                  Não sei            Pós graduação   
2419891            Pós graduação            Pós graduação   
1813163            Fundamental 1             Ensino Médio   
597158             Nunca Estudou        Educação Infantil   

                   Renda_familiar              Celular Computador Internet  \
1043297  De 1.212,01 ate 1.818,00              Sim, um        Não      Sim   
2355905  De 3.030,01 ate 3.636,00            Sim, dois    Sim, um      Sim   
2764872  De 1.212,01 ate 1.818,00            Sim, dois        Não      Sim   
1713255             Nenhuma Renda              Sim, um        Não      Não   
3038365  De 7.272,01 ate 8.484,00  Sim, quatro ou mais    Sim, um      Sim   
844591   De 2.424,01 ate 3.030,00            Sim, dois    Sim, um      Sim   
2330555  De 2.424,01 ate 3.030,00  Sim, quatro ou mais  Sim, dois      Sim   
2419891  De 1.818,01 ate 2.424,00            Sim, três  Sim, três      Sim   
1813163  De 1.818,01 ate 2.424,00              Sim, um    Sim, um      Sim   
597158               Ate 1.212,00            Sim, dois        Não      Sim   

         Nota_geral  
1043297         NaN  
2355905      650.34  
2764872         NaN  
1713255      452.54  
3038365      574.86  
844591          NaN  
2330555      512.12  
2419891      486.92  
1813163      529.56  
597158          NaN

Este dataframe agora contém 3476105 linhas e 22 colunas.

- Faixa_etaria: Intervalo de idades que agrupa pessoas com características similares. Tipo: Category.
-    Sexo: Gênero do candidato. Tipo: Category.
-    Estado_civil: Situação matrimonial do candidato. Tipo: Category.
-    Cor_raca: Classificação étnica ou racial. Tipo: Category.
-    St_conclusão: Situação de conclusão do aluno, por exemplo: Não Concluído, Já Concluído. Tipo: Category.
-    Ano_concluiu: Ano de conclusão do Ensino Médio. Tipo: Integer.
-    Escola: Tipo de escola (pública ou privada). Tipo: Category.
-    Treineiro: Indica se o aluno fez a prova para testar seus conhecimentos. Tipo: Category.
-    Sg_uf_esc: Unidade federativa onde o aluno realizou a prova. Tipo: Category.
-    Nota_cn: Nota obtida em Ciências da Natureza (Química, Biologia, Física). Tipo: Float.
-    Nota_lc: Nota obtida em Linguagens (Português, Literatura, Inglês). Tipo: Float.
-    Nota_mt: Nota obtida em Matemática. Tipo: Float.
-    Nota_redacao: Nota obtida na redação. Tipo: Float.
-    Nota_ch: Nota obtida em Ciências Humanas (História, Geografia, Filosofia). Tipo: Float.
-    Nota_geral: Média das notas anteriores. Tipo: Float.
-    Escolaridade_pai: Nível de educação da figura paterna. Tipo: Category.
-    Escolaridade_mae: Nível de educação da figura materna. Tipo: Category.
-    Moradores: Número de moradores na residência. Tipo: Integer.
-    Renda_familiar: Faixa de renda mensal da família. Tipo: Category.
-    Celular: Acesso ao celular e quantidade. Tipo: Category.
-    Computador: Acesso ao computador e quantidade. Tipo: Category.
-    Internet: Acesso à internet. Tipo: Category.


### 3.4 Manipulação para memória

In [38]:
t.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3476105 entries, 0 to 3476104
Data columns (total 19 columns):
 #   Column            Dtype  
---  ------            -----  
 0   Faixa_etaria      object 
 1   Sexo              object 
 2   Estado_civil      object 
 3   Cor_raca          object 
 4   Escola            object 
 5   Treineiro         object 
 6   Uf_esc            object 
 7   Nota_cn           float64
 8   Nota_ch           float64
 9   Nota_lc           float64
 10  Nota_mt           float64
 11  Nota_redacao      float64
 12  Escolaridade_pai  object 
 13  Escolaridade_mae  object 
 14  Renda_familiar    object 
 15  Celular           object 
 16  Computador        object 
 17  Internet          object 
 18  Nota_geral        float64
dtypes: float64(6), object(13)
memory usage: 2.8 GB


In [39]:
## Selecionando as colunas float
conv1 = t.select_dtypes('float64').columns

In [40]:
t[conv1] = t[conv1].astype('float32')

In [41]:
## Selecionando as colunas objeto
conv2 = t.select_dtypes('object').columns 

In [42]:
t[conv2] = t[conv2].astype('category')

Com a remoção das colunas desnecessárias e as alterações feitas até agora, já conseguimos reduzir o tamanho do dataframe de 7,0 GB para 3,1 GB. No próximo passo, iremos avaliar o impacto da mudança dos tipos de variáveis e verificar o uso de memória e exportar nossa dataframe para a análise.

## 4. Exportando

In [44]:
path = "C:\\Users\\Lucas\\Data Science P1\\Data Science//EnemProject.parquet"

In [46]:
t.to_parquet(path)

In [48]:
pd.read_parquet("C:\\Users\\Lucas\\Data Science P1\\Data Science//EnemProject.parquet").info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3476105 entries, 0 to 3476104
Data columns (total 19 columns):
 #   Column            Dtype   
---  ------            -----   
 0   Faixa_etaria      category
 1   Sexo              category
 2   Estado_civil      category
 3   Cor_raca          category
 4   Escola            category
 5   Treineiro         category
 6   Uf_esc            category
 7   Nota_cn           float32 
 8   Nota_ch           float32 
 9   Nota_lc           float32 
 10  Nota_mt           float32 
 11  Nota_redacao      float32 
 12  Escolaridade_pai  category
 13  Escolaridade_mae  category
 14  Renda_familiar    category
 15  Celular           category
 16  Computador        category
 17  Internet          category
 18  Nota_geral        float32 
dtypes: category(13), float32(6)
memory usage: 122.7 MB


Reduzimos nossa dataframe de 7.0GB para 128MB, agora nós poderemos manipular ela de uma forma mais fácil!